File: moderate/Predict-The-Number/Predict The Number.cs
using System;
using System.IO;
using System.Collections.Generic;

class Program
{
    static void Main(string[] args)
    {
        using (StreamReader reader = File.OpenText(args[0]))
        while (!reader.EndOfStream)
        {
            string line = reader.ReadLine();
            if (null == line)
                continue;
            ShowNumber(line);
        }
    }
    
    static void ShowNumber(string line){
        string result = "0";
        double index = Convert.ToInt64(line);
        if(index==0) Console.WriteLine(0);
            else if(index==1) Console.WriteLine(1);
            else Console.WriteLine(GetDigit(index,(int)Math.Log(index,2)+1));
    }
        
    static char GetDigit(double pos, int rowPos){
        if(rowPos>0){
            int dig = (int)Char.GetNumericValue(
                                GetDigit(pos-Math.Pow(2,rowPos-1),
                                        (int)Math.Log(pos-Math.Pow(2,rowPos-1),2)+1)
                                );
            return (dig!=2? (dig+1).ToString()[0] : '0' );  
        } else return '0';
    }
}

File: moderate/Predict-The-Number/Predict The Number.java
import java.io.*;
public class Main {
    public static void main (String[] args) throws IOException {
        File file = new File(args[0]);
        BufferedReader buffer = new BufferedReader(new FileReader(file));
        String line;
        while ((line = buffer.readLine()) != null) {
            line = line.trim();
            ShowNumber(line);
        }
    }
    
    static void ShowNumber(String line){
        String result = "0";
        double index = Double.valueOf(line);
        if(index==0) System.out.println(0);
            else if(index==1) System.out.println(1);
            else {
                int log2 = (int)((Math.log(index)/Math.log(2)));
                System.out.println(GetDigit(index,log2+1));
            } 
    }
        
    static char GetDigit(double pos, int rowPos){
        if(rowPos>0){
            int log2 = (int)(Math.log(pos-Math.pow(2,rowPos-1))/Math.log(2));
            char ch = GetDigit(pos-Math.pow(2,rowPos-1), log2+1);
            int dig = Integer.valueOf(Character.toString(ch));
            if(dig!=2) return Integer.toString(dig+1).charAt(0);
            else return '0';
        } else return '0';
    }
}
